test/shared/src/main/scala/zio/test/environment/package.scala

```diff
@@ -401,17 +342,100 @@ package object environment extends PlatformSpecific {
           case WarningData.Start          => ZIO.succeedNow(WarningData.done)
           case WarningData.Pending(fiber) => fiber.interrupt.as(WarningData.done)
         }
 
+      /**
+       * Polls until all descendants of this fiber are done or suspended.
+       */
+      private lazy val awaitSuspended: UIO[Unit] =
+        live.provide {
+          suspended.repeat {
+            Schedule.doUntilEquals(true) && Schedule.fixed(1.millisecond)
+          }
+        }.unit
+
+      /**
+       * Provides access to the list of descendants of this fiber (children and
+       * their children, recursively).
+       */
+      private def descendants: UIO[Iterable[Fiber.Runtime[Any, Any]]] =
+        for {
+          descriptor <- ZIO.descriptor
+          children   <- descriptor.children
+          collected  <- ZIO.foreach(children)(_.descendants)
+        } yield children ++ collected.flatten
+
+      /**
+       * Captures a "snapshot" of the status of all descendants of this fiber.
+       * Fails with the `Unit` value if any descendant of this fiber is not
+       * done or suspended. Note that because we cannot synchronize on the
+       * status of multiple fibers at the same time this snapshot may not be
+       * fully consistent.
+       */
+      private lazy val freeze: IO[Unit, Set[Fiber.Status]] =
+        descendants.flatMap { fibers =>
+          ZIO
+            .foreach(fibers)(_.status.filterOrFail {
+              case Fiber.Status.Done                     => true
+              case Fiber.Status.Suspended(_, _, _, _, _) => true
+              case _                                     => false
+            }(()))
+            .map(_.toSet)
+        }
+
+      /**
+       * Constructs a `Duration` from an `OffsetDateTime`.
+       */
+      private def fromDateTime(dateTime: OffsetDateTime): Duration =
+        Duration(dateTime.toInstant.toEpochMilli, TimeUnit.MILLISECONDS)
+
+      /**
+       * Runs all effects scheduled to occur on or before the specified
+       * duration, which may depend on the current time, in order.
+       */
+      private def run(f: Duration => Duration): UIO[Unit] =
+        awaitSuspended *>
+          clockState.modify { data =>
+            val end = f(data.duration)
+            data.sleeps.sortBy(_._1) match {
+              case (duration, promise) :: sleeps if duration <= end =>
+                (Some((end, promise)), Data(duration, sleeps, data.timeZone))
+              case _ => (None, Data(end, data.sleeps, data.timeZone))
+            }
+          }.flatMap {
+            case None => UIO.unit
+            case Some((end, promise)) =>
+              promise.succeed(()) *>
+                ZIO.yieldNow *>
+                run(_ => end)
+          }
+
+      /**
+       * Returns whether all descendants of this fiber are done or suspended.
+       */
+      private lazy val suspended: UIO[Boolean] =
+        freeze.zipWith(freeze)(_ == _).orElseSucceed(false)
```

Comment:
Is this just checking to see you get 2 successive reads with the same statuses? i.e. "stable statuses"?

Reply:
Yes.

Reply:
So there's a bit of a race condition here. What happens if it says that it is suspended, when it is not in fact suspended? Is there other logic to handle that case?

Reply:
Yes, we're relying on the two successive snapshots to address this, that if it isn't really done between the first snapshot and the second snapshot some fiber will have changed its status. This seems to be working in practice but I agree it isn't bulletproof. Not sure what the best solution is for this. I tried adding a delay between the snapshots but it seemed to cause issues on Scala.js. Maybe we yield between snapshots?

Reply:
Given the thread pools have FIFO delays, `yieldTo` would indeed be better. Although that would probably cause issues on Scala.js too (may be worth discovering what those are as they may point to another issue).

We should also maybe make a ticket for this race issue so we don't forget about it. I'm trying to think of what sort of tools would be necessary here.

Reply:
`yield` seemed to work for me locally on Scala.js. Running CI now.

Reply:
Okay, seems to be working. Hopefully this hardens it a little.

Reply:
Approved, and sorry for the delay! I think we should create a ticket for this one because it will eventually come back to haunt us...